pkg/controller/node/status.go

```diff
@@ -60,24 +89,93 @@ func calculateStatus(cconfig *v1.ControllerConfig, pool *mcfgv1.MachineConfigPoo
+	// ready == nodeResumed, updated
+	// unavailable == draining, cordoned
+	// degraded == if the condition.Reason == error
+	// this ensures that a MCP only enters Upgradeable==False if the node actually needs to upgrade to the new MC
+	for _, state := range mcs {
+		var ourNode *corev1.Node
+		for _, n := range nodes {
+			if state.Name == n.Name {
+				ourNode = n
+				break
+			}
+		}
+		if ourNode == nil {
+			klog.Errorf("Could not find specified node %s", state.Name)
+		}
+		if len(state.Status.Conditions) == 0 {
+			// not ready yet
+			break
+		}
+		for _, cond := range state.Status.Conditions {
+			if strings.Contains(cond.Message, "Error:") {
+				degradedMachines = append(degradedMachines, ourNode)
+				continue
+			}
+			if cond.Status == metav1.ConditionUnknown {
+				switch v1alpha1.StateProgress(cond.Type) {
+				case mcfgalphav1.MachineConfigNodeUpdatePrepared:
+					updatingMachines = append(updatedMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateExecuted:
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdatePostActionComplete:
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateComplete:
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeResumed:
+					updatedMachines = append(updatingMachines, ourNode)
+					readyMachines = append(readyMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateCompatible:
+					updatedMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateDrained:
+					unavailableMachines = append(unavailableMachines, ourNode)
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdateCordoned:
+					unavailableMachines = append(unavailableMachines, ourNode)
+					updatingMachines = append(updatingMachines, ourNode)
+				case mcfgalphav1.MachineConfigNodeUpdated:
+					updatedMachines = append(updatedMachines, ourNode)
+					readyMachines = append(readyMachines, ourNode)
+				}
+			}
+		}
+	}
+	degradedMachineCount := int32(len(degradedMachines))
 	updatedMachineCount := int32(len(updatedMachines))
-
-	readyMachines := getReadyMachines(pool, nodes)
+	unavailableMachineCount := int32(len(unavailableMachines))
+	updatingMachineCount := int32(len(updatingMachines))
 	readyMachineCount := int32(len(readyMachines))
 
-	unavailableMachines := getUnavailableMachines(nodes, pool)
-	unavailableMachineCount := int32(len(unavailableMachines))
+	// this is # 1 priority, get the upgrade states actually reporting
+	if degradedMachineCount+readyMachineCount+unavailableMachineCount+updatingMachineCount != int32(len(nodes)) {
+		klog.Infof("new state reporting did not get all nodes, falling back. Sate reporting node total %d and actual node total %d", (degradedMachineCount + readyMachineCount + updatedMachineCount + unavailableMachineCount + updatingMachineCount), len(nodes))
```

Comment:
question: for cases where the MCN is not enabled, I think these 2 messages will spam (so all 4.15 releases). See: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/4012/pull-ci-openshift-machine-config-operator-master-e2e-gcp-op/1729234371795226624/artifacts/e2e-gcp-op/gather-extra/artifacts/pods/openshift-machine-config-operator_machine-config-controller-54954c5cf8-lg5cq_machine-config-controller.log

I think we should remove these so as not to confuse the users, or gate them also behind the featuregate.

Like we talked above, let's make sure we create a card since this "fallback" is used by default and can be a bit wonky

Reply:
sure, i'll fix this log too